pandas/core/algorithms.py

```diff
@@ -1991,13 +1991,11 @@ def union_with_duplicates(lvals: ArrayLike, rvals: ArrayLike) -> ArrayLike:
     -----
     Caller is responsible for ensuring lvals.dtype == rvals.dtype.
     """
-    indexer = []
     l_count = value_counts(lvals, dropna=False)
     r_count = value_counts(rvals, dropna=False)
     l_count, r_count = l_count.align(r_count, fill_value=0)
+    final_count = np.maximum(l_count, r_count).astype("int", copy=False)
     unique_array = unique(concat_compat([lvals, rvals]))
     unique_array = ensure_wrapped_if_datetimelike(unique_array)
-
-    for i, value in enumerate(unique_array):
-        indexer += [i] * int(max(l_count.at[value], r_count.at[value]))
-    return unique_array.take(indexer)
+    repeats = final_count.reindex(unique_array).values  # type: ignore[attr-defined]
```

Comment:
Looks good,

can you refactor this to return the indexer and operate on the MultiIndex in _union()? This avoids losing the dtype.

Reply:
Refactored to preserve EA dtypes. The indexer applies to the combined unique values not the original MI's so I don't think we need to return it.